discovery.go

```diff
@@ -0,0 +1,369 @@
+				t.Reset(next)
+			case <-advertisingCtx.Done():
+				t.Stop()
+				return
+			}
+		}
+	}()
+}
+
+// StopAdvertise stops advertising this node's interest in a topic
+func (d *discover) StopAdvertise(topic string) {
+	if d.discovery == nil {
+		return
+	}
+
+	if advertiseCancel, ok := d.advertising[topic]; ok {
+		advertiseCancel()
+		delete(d.advertising, topic)
+	}
+}
+
+// Bootstrap handles initial bootstrapping of peers for a given topic
+func (d *discover) Bootstrap(async bool, topic string, dOpts ...discovery.Option) {
+	if d.discovery == nil {
+		return
+	}
+
+	d.bootstrappedMux.RLock()
+	doneCh, bootstrapStarted := d.bootstrapped[topic]
+	d.bootstrappedMux.RUnlock()
+
+	if bootstrapStarted {
+		if !async {
+			<-doneCh
+		}
+	} else {
+		d.bootstrappedMux.Lock()
+		doneCh, bootstrapStarted := d.bootstrapped[topic]
+		if bootstrapStarted {
+			d.bootstrappedMux.Unlock()
+			if !async {
+				<-doneCh
+			}
+			return
+		}
+		ch := make(chan struct{}, 1)
+		d.bootstrapped[topic] = ch
+		d.bootstrappedMux.Unlock()
+
+		if async {
+			go func() {
+				d.handleDiscovery(d.p.ctx, topic, dOpts)
+				ch <- struct{}{}
+				close(ch)
+			}()
+		} else {
+			d.handleDiscovery(d.p.ctx, topic, dOpts)
+			ch <- struct{}{}
+			close(ch)
+		}
+	}
+}
+
+// Discover searches for additional peers interested in a given topic
+func (d *discover) Discover(topic string, opts ...discovery.Option) {
+	if d.discovery == nil {
+		return
+	}
+
+	d.discoverQ <- &discoverReq{topic, opts}
+}
+
+func (d *discover) handleDiscovery(ctx context.Context, topic string, opts []discovery.Option) {
+	discoverCtx, cancel := context.WithTimeout(ctx, time.Second*10)
+	defer cancel()
+
+	peerCh, err := d.discovery.FindPeers(discoverCtx, topic, opts...)
+	if err != nil {
+		log.Debugf("error finding peers for topic %s: %v", topic, err)
+		return
+	}
+
+	findPeers := make(map[peer.ID]struct{})
+
+loop:
+	for {
+		select {
+		case pi, more := <-peerCh:
+			if pi.ID == d.p.host.ID() || pi.ID == "" {
+				if more {
+					continue
+				} else {
+					break loop
+				}
+			}
+
+			findPeers[pi.ID] = struct{}{}
+
+			go func(pi peer.AddrInfo) {
+				ctx, cancel := context.WithTimeout(ctx, time.Second*10)
+				defer cancel()
+
+				err := d.p.host.Connect(ctx, pi)
+				if err != nil {
+					log.Debugf("Error connecting to pubsub peer %s: %s", pi.ID, err.Error())
+					return
+				}
+			}(pi)
+
+			if !more {
+				break loop
+			}
+
+		case <-discoverCtx.Done():
+			log.Infof("pubsub discovery for %s timeout", topic)
+			return
+		}
+	}
+
+	if len(findPeers) > 0 {
+		req := &reqNewPeerNotifier{topic: topic, resp: make(chan chan peer.ID, 1)}
+		d.addPeerNotif <- req
+		peerCh := <-req.resp
+		defer func() { d.rmPeerNotif <- peerCh }()
+		for {
+			select {
+			case pid := <-peerCh:
+				if _, ok := findPeers[pid]; ok {
+					delete(findPeers, pid)
+					log.Debugf("Connected to pubsub peer %s", pid)
+					if len(findPeers) == 0 {
+						return
+					}
+				}
+			case <-discoverCtx.Done():
```

Comment:
We're going to wait the full timeout unless we magically manage to connect to every single peer.

Reply:
Anyways, it seems like we're tracking the wrong thing here. The goal is to find peers for the topic, not connect to all the peers that happened to be advertising the topic. What if we just kept a discovery worker running in the background till the pubsub router told us to turn it off? That is, `Discover(topic)` and `StopDiscover(topic)`. I'm not sure how well that would integrate with the current pubsub implementatoin but it would let us entirely drop the "new peer" logic.

Reply:
Note: this is non-blocking. While not perfect, this module should work well enough (although I'd like to consider this as I think it would simplify the logic here significantly).

Reply:
@Stebalien While I agree with you in principle that this is correct, any change to this behavior potentially breaks people using PubSub currently if they just copy-pasted the bootstrap code from go-libp2p-pubsub-router.

If we don't think this is happening, then I'm perfectly fine telling all the standard non-persistent messaging schemes to figure out the "ready condition" on their own (persistent schemes don't really care since the initially sent messages won't get dropped). It's possible that #190 could be used as events to trigger the "ready condition", but is there anything else we should expose?

@vyzo do you know how people are bootstrapping pubsub currently? Will changing behavior from go-libp2p-pubsub-router to make bootstrapping non-blocking be a problem?

Reply:
Also, we do not have to connect to every peer we just need to have a response from `connect` on every peer. For example, if connects fail by going to dead IP addresses then we'll get a Peer Dead notification which will also count as a completed attempt.

Reply:
Discussed in-person. The resolution was to:

1. Drop all the "wait to bootstrap" code.
2. Provide options to the `Publish` function that wait for peers.
3. Trigger a search for more peers on publish, if we don't have enough.

Reply:
@Stebalien How's this modified version?

1. Drop all the "wait to bootstrap" code.
2. _Provide options to a new `SetTopicParameters` function (@vyzo has expressed concern about breaking the `Publish` function in the past, but we could definitely do that instead if people prefer)._
3. Trigger a search for more peers on publish, if we don't have enough. _(block until we have enough as defined in 2)_
4. _Trigger a search for more peers on continuously, if we don't have enough. This helps with discovering new updates._

I haven't figured out if it makes sense to have the continuously search function built into the Routers (as we do currently in the PR for gossipsub) or passed in by the developer. I'm thinking the developer should define the behavior since some protocols (like floodsub) are super oblivious to network conditions and have no idea how many peers they want to have. However, I can certainly be convinced otherwise.

Reply:
> 1
:+1:

> 2

We could do that but it wouldn't be a _breaking_ change if we use a variable number of functional options (well, not that breaking).

> 3

Sgtm.

> 4

Regardless, we need a backoff on the continuous search. Probably something like:

1. Every time we search, double the delay up to some max.
2. Every time we loose a peer, reset this delay (if we're below the delay).

Given something like that, we can probably set the floodsub limit to something like 5.

Reply:
Although... if we block _at all_ inside publish, we really should have it take a context. Which will be a breaking change.

Honestly, we should seriously consider just making that change. Alternatively, we could just not do this here and, instead:

1. Add a that `Bootstrap(topic)` function.
2. Provide an external helper function `BootstrapWait(pubsub, topic, npeers)` and call it before calling publish. However, I'd _really_ prefer to make publish do this.

Reply:
Even if a little gross, that change doesn't need to necessarily be breaking since we could make the context specific to the bootstrap function passed into `Publish`. The result would probably be that after a timeout we send the message even if we don't think we're bootstrapped.

Reply:
Sure, but I'd _rather_ break things than do something like that.

Reply:
fair enough. @vyzo are you onboard?

Reply:
sure, i am fine with breaking things here.